### PR TITLE
[Projects] Fix BC for GET requests of non-normalized function names 

### PR DESCRIPTION
**Pull request by yaelgen**
follow up for: https://github.com/mlrun/mlrun/pull/3287
In this previous PR we added a normalization to a function name before saving it, and notify the user of the change.

We realized that in case our users ignored/missed the warning when setting the function, and then attempted to access (import or update) it later,  they would be unable to do so using the old name, and would not know why their function was not found.

As a result, as long as we do not block requests with non-normalized names, we will normalize GET requests as well.
In this fix we handle this case by adding fall-back to get_function:
Normalize the requested name and attempt to retrieve it from the database. if no response is received, we will check to see if the original name contained underscores,  if so, we will repeat the retrieval and return the result (if exists).

The backwards compatibility behavior here is:
In circumstances when the user has an old function with non-normalized name already saved in the database, he will be able to retrieve it.
If he makes an update request, we will create a new updated function, with a normalized name. Now if he attempts to retrieve it using the old name, we will first normalize the name in the request for him, and return the updated function.

### CODE REVIEW

File: tests/api/api/test_projects.py [hunk 1]
```diff
@@ -130,6 +131,54 @@ def test_redirection_from_worker_to_chief_delete_project(
                 assert response.text == expected_response
 
 
+def test_get_project_function_when_using_not_normalize_name(
```
Comment (Tankilevitch) [inline tests/api/api/test_projects.py:134]: Looking at the logic you are testing I think this need to be under the `tests_functions` and on the DB layer, as you are not actually testing any "API" logic over here but mainly that the `db` layer is able to return a function with an underscore.

File: tests/projects/test_project.py [hunk 2]
```diff
@@ -406,13 +406,55 @@ def test_set_function_requirements():
     ]
 
 
+def test_backwards_compatibility_get_non_normalized_function_name(rundb_mock):
+    project = mlrun.projects.MlrunProject(
+        "project", default_requirements=["pandas>1, <3"]
+    )
```
Comment (Tankilevitch) [inline tests/projects/test_project.py:412]:
```suggestion
    project = mlrun.projects.MlrunProject(
        "project"
    )
```

File: tests/projects/test_project.py [hunk 3]
```diff
@@ -406,13 +406,55 @@ def test_set_function_requirements():
...
+        handler="myhandler",
+        filename=func_path,
+    )
+    # nuclio also normalizes the name, so we de-normalize the function name before storing it
+    func.metadata.name = func_name
+
+    # mock the normalize function response in order to insert a non-normalized function name to the db
+    with unittest.mock.patch("mlrun.utils.normalize_name", return_value=func_name):
+        project.set_function(name=func_name, func=func)
+
```
Comment (Tankilevitch) [inline tests/projects/test_project.py:429]: 😍

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
Nice Job! very concise and straight to the point @yaelgen

**Review by liranbg — CHANGES_REQUESTED** [review]
Well done! few things
1. Add type hint / param explanation for public functions you encounter and change, will help future developers
2. Public functions before private functions, try to push down the internal implementation of get function after all public class apis
3. It is legitimate to log that you function was not found, trying to find function without normalizing its name

File: mlrun/api/db/sqldb/db.py [hunk 4]
```diff
@@ -984,6 +984,30 @@ def store_function(
...
+        try:
+            return self._get_function(
+                session, normalized_function_name, project, tag, hash_key
+            )
+        except mlrun.errors.MLRunNotFoundError as exc:
+            if "_" in name:
+                logger.debug(
+                    f"Function {normalized_function_name} was not found,"
+                    f" trying to find function by its original name"
+                )
```
Comment (liranbg) [inline mlrun/api/db/sqldb/db.py:1006]: Id change the debug verbose to warning
and change the log to use kwargs, so once we have json formatter, those kwargs would be indexed well

```python
logger.warning("Failed to get underscore-named function, trying without normalization", function_name=name)
```

**Review by liranbg — COMMENTED** [review]
last one

**Review by liranbg — APPROVED** [review]

File: mlrun/projects/project.py [hunk 5]
```diff
@@ -1639,21 +1641,48 @@ def get_function(
...
                 self, function, copy_function=copy_function
             )
             self.spec._function_objects[key] = function
+
         return function
 
+    def _get_function(self, key, sync, ignore_cache):
+        """
+        Function can be retrieved from the project spec (cache) or from the database.
+        In sync mode, we first preform a sync of the function_objects from the function_definitions,
```
Comment (Tankilevitch) [inline mlrun/projects/project.py:1664]:
```suggestion
        In sync mode, we first perform a sync of the function_objects from the function_definitions,
```

**Review by Tankilevitch — APPROVED** [review]
LGTM

**Pull request merged**